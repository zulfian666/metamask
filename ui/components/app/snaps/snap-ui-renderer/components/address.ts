import { AddressElement } from '@metamask/snaps-sdk/jsx';
import { UIComponentFactory } from './types';

export const address: UIComponentFactory<AddressElement> = ({ element }) => ({
  element: 'ConfirmInfoRowAddress',
  props: {
    address: element.props.address,
    isSnapUsingThis: true,
  },
});
