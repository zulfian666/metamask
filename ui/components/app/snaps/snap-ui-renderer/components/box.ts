import { BoxElement, JSXElement, BoxProps } from '@metamask/snaps-sdk/jsx';
import { getJsxChildren } from '@metamask/snaps-utils';
import { NonEmptyArray } from '@metamask/utils';
import {
  Display,
  FlexDirection,
  TextColor,
} from '../../../../../helpers/constants/design-system';
import { mapToTemplate } from '../utils';
import { UIComponent, UIComponentFactory } from './types';

function generateJustifyContent(alignment?: BoxProps['alignment']) {
  switch (alignment) {
    default:
    case 'start':
      return 'flex-start';

    case 'center':
      return 'center';

    case 'end':
      return 'flex-end';

    case 'space-between':
      return 'space-between';

    case 'space-around':
      return 'space-around';
  }
}

export const box: UIComponentFactory<BoxElement> = ({
  element,
  ...params
}) => ({
  element: 'Box',
  children: getJsxChildren(element).map((children) =>
    mapToTemplate({ ...params, element: children as JSXElement }),
  ) as NonEmptyArray<UIComponent>,
  props: {
    display: Display.Flex,
    flexDirection:
      element.props.direction === 'horizontal'
        ? FlexDirection.Row
        : FlexDirection.Column,
    justifyContent: generateJustifyContent(element.props.alignment),
    className: 'snap-ui-renderer__panel',
    color: TextColor.textDefault,
  },
});
