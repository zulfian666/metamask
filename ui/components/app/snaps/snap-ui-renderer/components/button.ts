import { ButtonElement } from '@metamask/snaps-sdk/jsx';
import { UIComponentFactory } from './types';

export const button: UIComponentFactory<ButtonElement> = ({ element }) => ({
  element: 'SnapUIButton',
  props: {
    type: element.props.type,
    variant: element.props.variant,
    name: element.props.name,
    disabled: element.props.disabled,
  },
  children: element.props.children,
});
