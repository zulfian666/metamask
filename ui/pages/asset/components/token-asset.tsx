import React, { useContext } from 'react';
import { Token } from '@metamask/assets-controllers';
import { useSelector, useDispatch } from 'react-redux';
import { getTokenTrackerLink } from '@metamask/etherscan-link';
import { useHistory } from 'react-router-dom';
import {
  getCurrentChainId,
  getRpcPrefsForCurrentProvider,
  getSelectedInternalAccount,
  getTokenList,
} from '../../../selectors';
import { isEqualCaseInsensitive } from '../../../../shared/modules/string-utils';
import { useTokenTracker } from '../../../hooks/useTokenTracker';
import { AssetType } from '../../../../shared/constants/transaction';
import { useTokenFiatAmount } from '../../../hooks/useTokenFiatAmount';
import {
  getURLHostName,
  roundToDecimalPlacesRemovingExtraZeroes,
} from '../../../helpers/utils/util';
import { MetaMetricsContext } from '../../../contexts/metametrics';
import { showModal } from '../../../store/actions';
import { MetaMetricsEventCategory } from '../../../../shared/constants/metametrics';
import AssetOptions from './asset-options';
import AssetPage from './asset-page';

const TokenAsset = ({ token }: { token: Token }) => {
  const { address, symbol, isERC721 } = token;

  const tokenList = useSelector(getTokenList);
  const chainId = useSelector(getCurrentChainId);
  const rpcPrefs = useSelector(getRpcPrefsForCurrentProvider);
  const { address: walletAddress } = useSelector(getSelectedInternalAccount);

  const history = useHistory();
  const dispatch = useDispatch();
  const trackEvent = useContext(MetaMetricsContext);

  const { name, iconUrl, aggregators } =
    Object.values(tokenList).find(
      (t) =>
        isEqualCaseInsensitive(t.symbol, symbol) &&
        isEqualCaseInsensitive(t.address, address),
    ) ?? {};

  const {
    tokensWithBalances,
  }: { tokensWithBalances: { string: string; balance: string }[] } =
    useTokenTracker({ tokens: [token], address: undefined });

  const balance = tokensWithBalances?.[0];
  const fiat = useTokenFiatAmount(address, balance?.string, symbol, {}, false);

  const tokenTrackerLink = getTokenTrackerLink(
    token.address,
    chainId,
    '',
    walletAddress,
    rpcPrefs,
  );

  return (
    <AssetPage
      asset={{
        chainId,
        type: AssetType.token,
        address,
        symbol,
        name,
        decimals: token.decimals,
        image: iconUrl,
        aggregators,
        balance: {
          value: balance?.balance,
          display: `${roundToDecimalPlacesRemovingExtraZeroes(
            balance?.string,
            5,
          )}`,
          fiat,
        },
        isERC721,
      }}
      optionsButton={
        <AssetOptions
          isNativeAsset={false}
          onRemove={() =>
            dispatch(
              showModal({ name: 'HIDE_TOKEN_CONFIRMATION', token, history }),
            )
          }
          onClickBlockExplorer={() => {
            trackEvent({
              event: 'Clicked Block Explorer Link',
              category: MetaMetricsEventCategory.Navigation,
              properties: {
                link_type: 'Token Tracker',
                action: 'Token Options',
                block_explorer_domain: getURLHostName(tokenTrackerLink),
              },
            });
            global.platform.openTab({ url: tokenTrackerLink });
          }}
          tokenSymbol={token.symbol}
        />
      }
    />
  );
};

export default TokenAsset;
