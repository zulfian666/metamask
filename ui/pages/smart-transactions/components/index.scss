.mm-smart-transactions-opt-in-modal {
  &__benefit {
    flex: 1;
  }

  &__icon {
    --size: 40px;
  }

  &__no-thanks-link {
    opacity: 75%;

    &:hover {
      color: var(--color-text-alternative) !important;
      text-decoration: none !important;
    }
  }

  .mm-modal-content {
    &__dialog {
      background-image: url('/images/smart-transactions/smart-transactions-opt-in-background.svg');
      background-position: -80px 16px;
      background-repeat: no-repeat;
    }
  }
}
