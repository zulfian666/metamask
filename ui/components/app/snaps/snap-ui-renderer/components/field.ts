import {
  FieldElement,
  InputElement,
  ButtonElement,
  JSXElement,
  DropdownElement,
} from '@metamask/snaps-sdk/jsx';
import { getJsxChildren } from '@metamask/snaps-utils';
import { button as buttonFn } from './button';
import { dropdown as dropdownFn } from './dropdown';
import { UIComponentFactory, UIComponentParams } from './types';

export const field: UIComponentFactory<FieldElement> = ({ element, form }) => {
  // For fields we don't render the Input itself, we just adapt SnapUIInput.
  const children = getJsxChildren(element);
  const child = children[0] as JSXElement;

  switch (child.type) {
    case 'Input': {
      const input = child as InputElement;
      const button = children[1] as ButtonElement;
      const buttonMapped =
        button &&
        buttonFn({ element: button } as UIComponentParams<ButtonElement>);
      return {
        element: 'SnapUIInput',
        props: {
          id: input.props.name,
          placeholder: input.props.placeholder,
          label: element.props.label,
          textFieldProps: {
            type: input.props.type,
          },
          name: input.props.name,
          form,
          error: element.props.error !== undefined,
          helpText: element.props.error,
        },
        propComponents: {
          endAccessory: buttonMapped && {
            ...buttonMapped,
            props: {
              ...buttonMapped.props,
              padding: 0,
            },
          },
        },
      };
    }

    case 'Dropdown': {
      const dropdown = child as DropdownElement;
      const dropdownMapped = dropdownFn({
        element: dropdown,
      } as UIComponentParams<DropdownElement>);
      return {
        element: 'SnapUIDropdown',
        props: {
          ...dropdownMapped.props,
          id: dropdown.props.name,
          label: element.props.label,
          name: dropdown.props.name,
          form,
          error: element.props.error,
        },
      };
    }

    default:
      throw new Error(`Invalid Field child: ${child.type}`);
  }
};
