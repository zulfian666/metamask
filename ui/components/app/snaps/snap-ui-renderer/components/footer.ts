import { FooterElement, ButtonElement } from '@metamask/snaps-sdk/jsx';
import { getJsxChildren } from '@metamask/snaps-utils';

import {
  BackgroundColor,
  BlockSize,
  Display,
  FlexDirection,
} from '../../../../../helpers/constants/design-system';
import { UIComponentFactory, UIComponentParams } from './types';
import { button as buttonFn } from './button';

export const DEFAULT_FOOTER = {
  element: 'Box',
  key: 'default-footer',
  props: {
    display: Display.Flex,
    flexDirection: FlexDirection.Row,
    width: BlockSize.Full,
    padding: 4,
    className: 'snap-ui-renderer__footer',
    backgroundColor: BackgroundColor.backgroundDefault,
    style: {
      boxShadow: 'var(--shadow-size-md) var(--color-shadow-default)',
      height: '74px',
      position: 'fixed',
      bottom: 0,
    },
  },
};

const getDefaultButtons = (
  footer: FooterElement,
  t: (value: string) => string,
  onCancel?: () => void,
) => {
  const children = getJsxChildren(footer);

  // If onCancel is omitted by the caller we assume that it is safe to not display the default footer.
  if (children.length === 1 && onCancel) {
    return {
      element: 'SnapFooterButton',
      key: 'default-button',
      props: {
        onCancel,
        isSnapAction: false,
      },
      children: t('cancel'),
    };
  }

  return undefined;
};

export const footer: UIComponentFactory<FooterElement> = ({
  element,
  t,
  onCancel,
  ...params
}) => {
  const defaultButtons = getDefaultButtons(element, t, onCancel);

  const footerChildren = (getJsxChildren(element) as ButtonElement[]).map(
    (children, index) => {
      const buttonMapped = buttonFn({
        ...params,
        element: children,
      } as UIComponentParams<ButtonElement>);
      return {
        element: 'SnapFooterButton',
        key: `snap-footer-button-${buttonMapped.props?.name ?? index}`,
        props: {
          ...buttonMapped.props,
          isSnapAction: true,
        },
        children: buttonMapped.children,
      };
    },
  );

  if (defaultButtons) {
    footerChildren.unshift(defaultButtons);
  }

  return {
    ...DEFAULT_FOOTER,
    children: footerChildren,
  };
};
