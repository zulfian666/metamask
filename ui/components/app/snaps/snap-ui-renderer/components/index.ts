import { box } from './box';
import { heading } from './heading';
import { text } from './text';
import { divider } from './divider';
import { spinner } from './spinner';
import { image } from './image';
import { row } from './row';
import { address } from './address';
import { copyable } from './copyable';
import { button } from './button';
import { form } from './form';
import { input } from './input';
import { bold } from './bold';
import { italic } from './italic';
import { link } from './link';
import { field } from './field';
import { dropdown } from './dropdown';
import { value } from './value';

export const COMPONENT_MAPPING = {
  Box: box,
  Heading: heading,
  Text: text,
  Divider: divider,
  Spinner: spinner,
  Image: image,
  Copyable: copyable,
  Row: row,
  Address: address,
  Button: button,
  Form: form,
  Input: input,
  Bold: bold,
  Italic: italic,
  Link: link,
  Field: field,
  Dropdown: dropdown,
  Value: value,
};
