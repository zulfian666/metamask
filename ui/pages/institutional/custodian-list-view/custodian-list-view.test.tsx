import React from 'react';
import { render, waitFor } from '@testing-library/react';
import { Box } from '../../../components/component-library';
import CustodianListView from './custodian-list-view';

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useHistory: jest.fn(() => []),
}));

describe('CustodianListView', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should render CustodianListView', async () => {
    const mockCustodianList = [
      <Box>Custodian 1</Box>,
      <Box>Custodian 2</Box>,
      <Box>Custodian 3</Box>,
    ];
    const { container, getByText } = render(
      <CustodianListView custodianList={mockCustodianList} />,
    );
    expect(container).toMatchSnapshot();

    await waitFor(() => {
      expect(getByText('Custodian 1')).toBeInTheDocument();
      expect(getByText('Custodian 2')).toBeInTheDocument();
      expect(getByText('Custodian 3')).toBeInTheDocument();
    });
  });
});
