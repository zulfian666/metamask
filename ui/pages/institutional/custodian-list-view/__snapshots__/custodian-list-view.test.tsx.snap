// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustodianListView should render CustodianListView 1`] = `
<div>
  <div
    class="mm-box mm-box--padding-4 mm-box--display-flex mm-box--flex-direction-column mm-box--width-full"
    data-testid="connect-custodial-account"
  >
    <div
      class="mm-box mm-box--margin-top-4 mm-box--margin-bottom-4 mm-box--display-flex mm-box--align-items-center"
    >
      <button
        aria-label="[back]"
        class="mm-box mm-button-icon mm-button-icon--size-sm mm-box--display-flex mm-box--justify-content-center mm-box--align-items-center mm-box--color-icon-default mm-box--background-color-transparent mm-box--rounded-lg"
      >
        <span
          class="mm-box mm-icon mm-icon--size-sm mm-box--display-inline-block mm-box--color-inherit"
          style="mask-image: url('./images/icons/arrow-left.svg');"
        />
      </button>
      <p
        class="mm-box mm-text mm-text--body-md mm-box--color-text-default"
      >
        [back]
      </p>
    </div>
    <h4
      class="mm-box mm-text mm-text--body-lg-medium mm-box--margin-top-4 mm-box--color-text-default"
    >
      [connectCustodialAccountTitle]
    </h4>
    <h6
      class="mm-box mm-text mm-text--body-md mm-box--margin-top-2 mm-box--margin-bottom-5 mm-box--color-text-default"
    >
      [connectCustodialAccountMsg]
    </h6>
    <div
      class="mm-box"
    >
      <ul
        class="mm-box mm-box--width-full"
      >
        <div
          class="mm-box"
        >
          Custodian 1
        </div>
        <div
          class="mm-box"
        >
          Custodian 2
        </div>
        <div
          class="mm-box"
        >
          Custodian 3
        </div>
      </ul>
    </div>
  </div>
</div>
`;
